packages/muelu/src/Transfers/GeneralGeometric/MueLu_GeneralGeometricPFactory_def.hpp

```diff
@@ -0,0 +1,843 @@
+      // or the factory declared to provide "Coordinates".
+      // Then we check if the coordinates are available and eventually declare them as input.
+      RCP<const FactoryBase> myCoordsFact = GetFactory("Coordinates");
+      if (myCoordsFact == Teuchos::null) { myCoordsFact = fineLevel.GetFactoryManager()->GetFactory("Coordinates"); }
+      if (fineLevel.IsAvailable("Coordinates", myCoordsFact.get())) {
+        fineLevel.DeclareInput("Coordinates", myCoordsFact.get(), this);
+      }
+    }
+  }
+
+  template <class Scalar,class LocalOrdinal, class GlobalOrdinal, class Node>
+  void GeneralGeometricPFactory<Scalar, LocalOrdinal, GlobalOrdinal, Node>::Build(Level& fineLevel, Level& coarseLevel) const {
+    return BuildP(fineLevel, coarseLevel);
+  }
+
+  template <class Scalar,class LocalOrdinal, class GlobalOrdinal, class Node>
+  void GeneralGeometricPFactory<Scalar, LocalOrdinal, GlobalOrdinal, Node>::BuildP(Level& fineLevel, Level& coarseLevel) const {
+    FactoryMonitor m(*this, "Build", coarseLevel);
+
+    // obtain general variables
+    RCP<Matrix>      A             = Get< RCP<Matrix> >      (fineLevel, "A");
+    RCP<MultiVector> fineNullspace = Get< RCP<MultiVector> > (fineLevel, "Nullspace");
+
+    // get user-provided coarsening rate parameter (constant over all levels)
+    const ParameterList& pL = GetParameterList();
+
+    // collect general input data
+    LO blkSize            = A->GetFixedBlockSize();
+    RCP<const Map> rowMap = A->getRowMap();
+    LO ndofs              = rowMap->getNodeNumElements();
+    LO nNodes             = ndofs/blkSize;
+    LO ndm                = 0;
+    LO nFpts              = 0;
+
+    // RCP<Teuchos::FancyOStream> fancy = Teuchos::fancyOStream(Teuchos::rcpFromRef(std::cout));
+    // Teuchos::FancyOStream& out = *fancy;
+    typedef Xpetra::MultiVector<double,LO,GO,NO> xdMV;
+    RCP<xdMV> fineCoords = Teuchos::null;
+    RCP<xdMV> coarseCoords;
+    if (fineLevel.GetLevelID() == 0 &&
+	fineLevel.IsAvailable("Coordinates", NoFactory::get())) {
+      fineCoords = fineLevel.Get< RCP<xdMV> >("Coordinates", NoFactory::get());
+      ndm = fineCoords->getNumVectors();
+      nFpts = fineCoords->getGlobalLength();
+    } else {
+      RCP<const FactoryBase> myCoordsFact = GetFactory("Coordinates");
+      if (myCoordsFact == Teuchos::null) { myCoordsFact = fineLevel.GetFactoryManager()->GetFactory("Coordinates"); }
+      if (fineLevel.IsAvailable("Coordinates", myCoordsFact.get())) {
+	fineCoords = fineLevel.Get< RCP<xdMV> >("Coordinates", myCoordsFact.get());
+	ndm = fineCoords->getNumVectors();
+	nFpts = fineCoords->getGlobalLength();
+      }
+    }
+    TEUCHOS_TEST_FOR_EXCEPTION(fineCoords==Teuchos::null, Exceptions::RuntimeError, "Coordinates cannot be accessed from fine level!");
+      
+    
+    // Get the coarsening rate
+    std::string CoarsenRate = pL.get<std::string>("Coarsen");
+    ArrayRCP<LO> coarseRate=arcp<LO>(3);
```

Comment:
Actually the algorithm is checking if the simulation is 1D, 2D or 3D but coarseRate is always of size 3 and some coarsening rates are specified as 1 if ndm<3. This allows me to avoid multiplying if statements later in the code.